Storage/storage.h

```diff
@@ -249,80 +274,93 @@ class GenericStorageImpl {
   }
 
-  template <typename F1, typename F2>
+  template <current::locks::MutexLockStatus MLS = current::locks::MutexLockStatus::NeedToLock, typename F1, typename F2>
   ::current::Future<::current::storage::TransactionResult<void>, ::current::StrictFuture::Strict> ReadOnlyTransaction(
       F1&& f1, F2&& f2) const {
-    return transaction_policy_.Transaction([&f1, this]() { return f1(static_cast<const FIELDS&>(fields_)); },
-                                           std::forward<F2>(f2));
+    current::locks::SmartMutexLockGuard<MLS> lock(persister_.Stream()->Impl()->publishing_mutex);
+    return transaction_policy_.TransactionFromLockedSection(
+        [&f1, this]() { return f1(static_cast<const FIELDS&>(fields_)); }, std::forward<F2>(f2));
   }
 
   void ExposeRawLogViaHTTP(int port, const std::string& route) { persister_.ExposeRawLogViaHTTP(port, route); }
 
-  typename std::result_of<decltype(&persister_t::InternalExposeStream)(persister_t)>::type InternalExposeStream() {
-    return persister_.InternalExposeStream();
-  }
-
-  persister_t& Persister() { return persister_; }
+  Borrowed<stream_t> BorrowUnderlyingStream() const { return persister_.BorrowStream(); }
+  const WeakBorrowed<stream_t>& UnderlyingStream() const { return persister_.Stream(); }
 
-  uint64_t TransactionsCount() const { return transactions_count_.GetValue(); }
+  // TODO(dkorolev): Custom publisher that refuses to publish transactions!
+  Borrowed<typename stream_t::publisher_t> PublisherUsed() { return persister_.PublisherUsed(); }
 
-  void WaitForTransactionsCount(uint64_t count) const {
-    transactions_count_.Wait([count](uint64_t value) { return value >= count; });
+  template <typename TYPE_SUBSCRIBED_TO = typename stream_t::entry_t, typename F>
+  typename stream_t::template SubscriberScope<F, TYPE_SUBSCRIBED_TO> Subscribe(
+      F& subscriber, uint64_t begin_idx = 0u, std::function<void()> done_callback = nullptr) const {
+    return persister_.Stream()->template Subscribe<TYPE_SUBSCRIBED_TO, F>(subscriber, begin_idx, done_callback);
   }
 
+  // Note: This method must be called from the stream-publishing-mutex-unlocked section,
+  // as terminating the transactions-replaying stream subscription thread from the following storage
+  // neeeds to lock that mutex from the destructor when de-registering its stream subscriber.
+  // TODO(dkorolev): Revisit the Sherlock-related logic here. Flip Sherlock first!
   void FlipToMaster() {
-    std::lock_guard<std::mutex> lock(mutex_);
-    if (role_ == StorageRole::Follower) {
-      persister_.template AcquireDataAuthority<current::locks::MutexLockStatus::AlreadyLocked>();
-      role_ = StorageRole::Master;
-    } else {
-      CURRENT_THROW(StorageIsAlreadyMasterException());
-    }
+    // TODO(dkorolev): Lock the mutex.
+    // TODO(dkorolev): Start from the stream.
+    persister_.BecomeMasterStorage();
   }
 
   void GracefulShutdown() { transaction_policy_.GracefulShutdown(); }
```

Comment:
Let's take this one offline. The value of `GracefulShutdown` is vague for me as of right now, as it appear to not make sense outside the usecases of "prepare to make the storage read-only (perhaps temporarily because a master flip is being prepared") and "stop serving storage data via HTTP as the storage is being destructed".